Perl/Decoder/t/091_thaw_order.t

```diff
@@ -0,0 +1,127 @@
+            },
+        }, $class;
+    }
+
+
+    sub FREEZE {
+        my ( $self, $serializer ) = @_;
+
+        push @steps, 'FooOutside::FREEZE(' . ::dd($self) . ')';
+        return {
+            attrs  => $self->{attrs},
+            inside => [
+                map { $self->{inside}{$_} } sort { $a <=> $b } keys %{$self->{inside}},
+            ],
+        };
+    }
+
+    sub THAW {
+        my ( $class, $serializer, @data ) = @_;
+
+        push @steps, 'FooOutside::THAW(' . ::dd($data[0]) . ')';
+        return $class->new( $data[0]->{attrs}, @{ $data[0]->{inside} // [] } );
+    }
+}
+
+my $struct = 
+    FooOutside->new(
+        # Attrs
+        {
+            attr1 => 'foobar',
+            bar => FooInside->new({}, 'should_be_first'),
+        },
+        FooInside->new({}, 'second/not_frozen', 'third/not_frozen'),
+        FooOutside->new({}, FooInside->new({}, 'fourth/inner')),
+    );
+
+my $srl = Sereal::Encoder::encode_sereal($struct, { freeze_callbacks => 1, canonical => 1 });
+my $dec = decode_sereal($srl);
+
+
+is_deeply(\@steps, [
+          "FooOutside::FREEZE(bless( {'attrs' => {'attr1' => 'foobar','bar' => bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' )},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ),'1' => bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' )}}, 'FooOutside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ))",
+          "FooOutside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' ))",
+          "FooInside::THAW({'attrs' => {},'inside' => ['fourth/inner']})",
+          "FooOutside::THAW({'attrs' => {},'inside' => [bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )]})",
+          "FooInside::THAW({'attrs' => {},'inside' => ['second/not_frozen','third/not_frozen']})",
+          "FooInside::THAW({'attrs' => {},'inside' => ['should_be_first']})",
+          "FooOutside::THAW({'attrs' => {'attr1' => 'foobar','bar' => bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' )},'inside' => [bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ),bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' )]})"
+        ], "freeze/thaw sequence for nested objects was as expected") or do {
+    local $Data::Dumper::Useqq = 1;
+    print Dumper(\@steps),"\n";
```

Comment:
I would expect `diag` here, not `print`.

(Also, I'm not a big fan of hardcoding Data::Dumper output to compare against because if Data::Dumper changes only a single character in its output format, the whole test will fail.)

Reply:
> I would expect diag here, not print.

Ah. diag does some of its own escaping and modifications, which I *dont* want. Ill fix this.

> Data::Dumper changes only a single character in its output format

Data::Dumper changes very rarely, and this output doesn't use anything particularly interesting. I can live with this.